src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -35,11 +36,229 @@ trait EntityComponent {
+    }
+
+    def rename(workspaceContext: SlickWorkspaceContext, entityType: String, oldName: String, newName: String): ReadWriteAction[Int] = {
+      findEntityByName(workspaceContext.workspaceId, entityType, oldName).map(_.name).update(newName)
+    }
+
+    def getEntityTypes(workspaceContext: SlickWorkspaceContext): ReadAction[TraversableOnce[String]] = {
+      filter(_.workspaceId === workspaceContext.workspaceId).map(_.entityType).distinct.result
+    }
 
+    def listEntitiesAllTypes(workspaceContext: SlickWorkspaceContext): ReadAction[TraversableOnce[Entity]] = {
+      // query gets entity records, all of its attributes (if there are any), and referred entities (if there are any)
+      val entityAndAttributesQuery =
+        findEntityByWorkspace(workspaceContext.workspaceId) joinLeft entityAttributesWithRefsJoin on (_.id === _._1._1.entityId)
+
+      unmarshalEntities(entityAndAttributesQuery)
+    }
+
+    def cloneAllEntities(sourceWorkspaceContext: SlickWorkspaceContext, destWorkspaceContext: SlickWorkspaceContext): ReadWriteAction[Unit] = {
+      val allEntitiesAction = listEntitiesAllTypes(sourceWorkspaceContext)
+
+      allEntitiesAction.flatMap(cloneEntities(destWorkspaceContext, _))
+    }
+
+    def cloneEntities(destWorkspaceContext: SlickWorkspaceContext, entities: TraversableOnce[Entity]): ReadWriteAction[Unit] = {
+      //1. First save JUST the entity record for the entity without any attributes (references)
+      val insertEntitiesOnly = DBIO.seq(entities.map(entity => save(destWorkspaceContext, entity.copy(attributes = Map.empty))).toSeq:_*)
+
+      //2. Re-save the entities, but this time with the attributes that were previously omitted
+      //Since all entities were already saved in step 1, we can now save the entities again with their references to other entities
+      val insertEntitiesAndAttributes = DBIO.seq(entities.map(entity => save(destWorkspaceContext, entity)).toSeq:_*)
+
+      insertEntitiesOnly andThen insertEntitiesAndAttributes
+    }
+
+    /**
+     * Starting with entities specified by entityIds, recursively walk down references accumulating all the ids
+     * @param entityIds the ids to start with
+     * @param accumulatedIds the ids accumulated from the prior call. If you wish entityIds to be in the overall
+     *                       results, start with entityIds == accumulatedIds, otherwise start with Seq.empty but note
+     *                       that if there is a cycle some of entityIds may be in the result anyway
+     * @return the ids of all the entities referred to by entityIds
+     */
+    private def recursiveGetEntityReferenceIds(entityIds: Set[Long], accumulatedIds: Set[Long]): ReadAction[Set[Long]] = {
+      // need to batch because some RDBMSes have a limit on the length of an in clause
+      val batchedEntityIds = createBatches(entityIds)
+
+      val batchQueries = batchedEntityIds.map {
+        idBatch => filter(_.id inSetBind(idBatch)) join
```

Comment:
uses bind parameters instead of embedding the strings directly in the query, avoids sql injection attacks among other things.